src/stream.rs

```diff
@@ -3,102 +3,57 @@
 //!  There is no dependency on actual TLS implementations. Everything like
 //! `native_tls` or `openssl` will work as long as there is a TLS stream supporting standard
 //! `Read + Write` traits.
+use std::pin::Pin;
+use std::task::{Context, Poll};
 
-use std::io::{Error as IoError, Read, Result as IoResult, Write};
-use std::net::SocketAddr;
-
-use bytes::{Buf, BufMut};
-use futures::Poll;
+use pin_project::pin_project;
 use tokio_io::{AsyncRead, AsyncWrite};
 
-/// Trait to switch TCP_NODELAY.
-pub trait NoDelay {
-    /// Set the TCP_NODELAY option to the given value.
-    fn set_nodelay(&mut self, nodelay: bool) -> IoResult<()>;
-}
-
-/// Trait to get the remote address from the underlying stream.
-pub trait PeerAddr {
-    /// Returns the remote address that this stream is connected to.
-    fn peer_addr(&self) -> IoResult<SocketAddr>;
-}
-
 /// Stream, either plain TCP or TLS.
+#[pin_project]
 pub enum Stream<S, T> {
     /// Unencrypted socket stream.
     Plain(S),
     /// Encrypted socket stream.
     Tls(T),
 }
 
-impl<S: Read, T: Read> Read for Stream<S, T> {
-    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
-        match *self {
-            Stream::Plain(ref mut s) => s.read(buf),
-            Stream::Tls(ref mut s) => s.read(buf),
-        }
-    }
-}
-
-impl<S: Write, T: Write> Write for Stream<S, T> {
-    fn write(&mut self, buf: &[u8]) -> IoResult<usize> {
-        match *self {
-            Stream::Plain(ref mut s) => s.write(buf),
-            Stream::Tls(ref mut s) => s.write(buf),
-        }
-    }
-    fn flush(&mut self) -> IoResult<()> {
-        match *self {
-            Stream::Plain(ref mut s) => s.flush(),
-            Stream::Tls(ref mut s) => s.flush(),
-        }
-    }
-}
-
-impl<S: NoDelay, T: NoDelay> NoDelay for Stream<S, T> {
-    fn set_nodelay(&mut self, nodelay: bool) -> IoResult<()> {
-        match *self {
-            Stream::Plain(ref mut s) => s.set_nodelay(nodelay),
-            Stream::Tls(ref mut s) => s.set_nodelay(nodelay),
+impl<S: AsyncRead, T: AsyncRead> AsyncRead for Stream<S, T> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::io::Result<usize>> {
+        match self.project() {
```

Comment:
This is a project type, so may have to just match deref rather than pin project, but that requires creating a pin afterwards. map_unchecked_mut might be better?